diff --git a/headers/private/kernel/lock.h b/headers/private/kernel/lock.h
index 05a4a4f77a..daeb9c55cf 100644
--- a/headers/private/kernel/lock.h
+++ b/headers/private/kernel/lock.h
@@ -27,6 +27,7 @@ typedef struct mutex {
 #else
 	int32					count;
 #endif
+	bigtime_t				wait_time;
 	uint8					flags;
 } mutex;
 
@@ -99,7 +100,7 @@ typedef struct rw_lock {
 // static initializers
 #if KDEBUG
 #	define MUTEX_INITIALIZER(name) \
-	{ name, NULL, B_SPINLOCK_INITIALIZER, -1, 0 }
+	{ name, NULL, B_SPINLOCK_INITIALIZER, -1, 0, 0 }
 #	define RECURSIVE_LOCK_INITIALIZER(name)	{ MUTEX_INITIALIZER(name), 0 }
 #else
 #	define MUTEX_INITIALIZER(name) \
diff --git a/src/system/kernel/locks/lock.cpp b/src/system/kernel/locks/lock.cpp
index 067b73d08b..fb08865137 100644
--- a/src/system/kernel/locks/lock.cpp
+++ b/src/system/kernel/locks/lock.cpp
@@ -39,6 +39,7 @@ struct mutex_waiter {
 	Thread*			thread;
 	mutex_waiter*	next;		// next in queue
 	mutex_waiter*	last;		// last in queue (valid for the first in queue)
+	bigtime_t		start;
 };
 
 struct rw_lock_waiter {
@@ -861,6 +862,7 @@ mutex_init_etc(mutex* lock, const char *name, uint32 flags)
 #else
 	lock->count = 0;
 #endif
+	lock->wait_time = 0;
 	lock->flags = flags & MUTEX_FLAG_CLONE_NAME;
 
 	T_SCHEDULING_ANALYSIS(InitMutex(lock, name));
@@ -1003,6 +1005,20 @@ _mutex_lock(mutex* lock, void* _locker)
 	} else if (lock->holder == 0) {
 		panic("_mutex_lock(): using uninitialized lock %p", lock);
 	}
+
+#	if KDEBUG_DETECT_MUTEX_DEADLOCKS
+	Thread* thread = Thread::Get(lock->holder);
+	if (thread != NULL) {
+		if (thread_is_blocked(thread) && thread->wait.type == THREAD_BLOCK_TYPE_MUTEX) {
+			const mutex* otherMutex = reinterpret_cast<const mutex*>(thread->wait.object);
+			if (otherMutex->holder == thread_get_current_thread_id()) {
+				panic("_mutex_lock() deadlock: attempted lock of %p, but holder is waiting on %p",
+					lock, otherMutex);
+			}
+		}
+		thread->ReleaseReference();
+	}
+#	endif
 #else
 	if ((lock->flags & MUTEX_FLAG_RELEASED) != 0) {
 		lock->flags &= ~MUTEX_FLAG_RELEASED;
@@ -1012,6 +1028,7 @@ _mutex_lock(mutex* lock, void* _locker)
 
 	// enqueue in waiter list
 	mutex_waiter waiter;
+	waiter.start = system_time();
 	waiter.thread = thread_get_current_thread();
 	waiter.next = NULL;
 
@@ -1068,6 +1085,17 @@ _mutex_unlock(mutex* lock)
 		lock->holder = waiter->thread->id;
 #endif
 
+#if 0
+		bigtime_t waited = system_time() - waiter->start;
+		if (waited > 1*1000*1000)
+			panic("thread %d waited on mutex %p for over one second", waiter->thread->id, lock);
+		lock->wait_time += waited;
+		if (lock->wait_time > 10*1000*1000) {
+			panic("mutex %p cumulative wait over 10 seconds", lock);
+			lock->wait_time = 0;
+		}
+#endif
+
 		// unblock thread
 		thread_unblock(waiter->thread, B_OK);
 	} else {
@@ -1114,6 +1142,7 @@ _mutex_lock_with_timeout(mutex* lock, uint32 timeoutFlags, bigtime_t timeout)
 
 	// enqueue in waiter list
 	mutex_waiter waiter;
+	waiter.start = system_time();
 	waiter.thread = thread_get_current_thread();
 	waiter.next = NULL;
 
diff --git a/src/system/kernel/smp.cpp b/src/system/kernel/smp.cpp
index 161ed52ff6..f5095c5b33 100644
--- a/src/system/kernel/smp.cpp
+++ b/src/system/kernel/smp.cpp
@@ -192,6 +192,13 @@ update_lock_contention(spinlock* lock, bigtime_t start)
 	const bigtime_t now = system_time();
 	lock->last_acquired = now;
 	lock->total_wait += (now - start);
+
+#if 0
+	if (lock->total_wait > 100*1000 && !debug_debugger_running()) {
+		panic("spinlock %p was waited on for 100ms\n", lock);
+		lock->total_wait=0;
+	}
+#endif
 }
 
 
